textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/regexmatch/GramBooleanQuery.java

```diff
@@ -293,6 +294,133 @@ private static String toLuceneQueryString(GramBooleanQuery query) {
+		if (left.isEmpty()) {
+			return right;
+		}
+		if (right.isEmpty()) {
+			return left;
+		}
+		GramBooleanQuery resultQuery = new GramBooleanQuery(QueryOp.OR);
+		for (String leftOperand : left.operandSet) {
+			for (String rightOperand : right.operandSet) {
+				GramBooleanQuery tempQuery = new GramBooleanQuery(QueryOp.AND);
+				tempQuery.operandSet.add(leftOperand);
+				tempQuery.operandSet.add(rightOperand);
+				resultQuery.subQuerySet.add(tempQuery);
+			}
+			for (GramBooleanQuery rightSubQuery : right.subQuerySet) {
+				GramBooleanQuery tempQuery = new GramBooleanQuery(QueryOp.AND);
+				tempQuery.operandSet.add(leftOperand);
+				tempQuery.operandSet.addAll(rightSubQuery.operandSet);
+				resultQuery.subQuerySet.add(tempQuery);
+			}
+		}
+		for (GramBooleanQuery leftSubQuery : left.subQuerySet) {
+			for (String rightOperand : right.operandSet) {
+				GramBooleanQuery tempQuery = new GramBooleanQuery(QueryOp.AND);
+				tempQuery.operandSet.addAll(leftSubQuery.operandSet);
+				tempQuery.operandSet.add(rightOperand);
+				resultQuery.subQuerySet.add(tempQuery);
+			}
+			for (GramBooleanQuery rightSubQuery : right.subQuerySet) {
+				GramBooleanQuery tempQuery = new GramBooleanQuery(QueryOp.AND);
+				tempQuery.operandSet.addAll(leftSubQuery.operandSet);
+				tempQuery.operandSet.addAll(rightSubQuery.operandSet);
+				resultQuery.subQuerySet.add(tempQuery);
+			}
+		}
+		return resultQuery;
+	}
+	
+	// After Transforming to DNF, apply Absorption laws to simplify it
+	// a OR (a AND b) --> a
+	// Tree must be already transformed to DNF before calling this function!
+	public static GramBooleanQuery simplifyDNF(GramBooleanQuery query) {
+		GramBooleanQuery result = new GramBooleanQuery(QueryOp.OR);
+		result.operandSet.addAll(query.operandSet);
+		
+		
+		Iterator<GramBooleanQuery> outerIterator = query.subQuerySet.iterator();
+		OuterLoop:
+		while (outerIterator.hasNext()) {
+			GramBooleanQuery outerAndQuery = outerIterator.next();
+			for (String operand : query.operandSet) {
+				if (outerAndQuery.operandSet.contains(operand)) {
+					continue OuterLoop;
+				}
+			}
+			Iterator<GramBooleanQuery> innerIterator = query.subQuerySet.iterator();
+			while (innerIterator.hasNext()) {
+				GramBooleanQuery innerAndQuery = innerIterator.next();
+				if (outerAndQuery != innerAndQuery) {
+					if (outerAndQuery.operandSet.containsAll(innerAndQuery.operandSet)) {
+						outerIterator.remove();
+						continue OuterLoop;
+					}				
+				}
+			}
+			// if reach this code, then add a copy of it to result
+			GramBooleanQuery tempQuery = new GramBooleanQuery(QueryOp.AND);
+			tempQuery.operandSet.addAll(outerAndQuery.operandSet);
+			result.subQuerySet.add(tempQuery);
+		}
+		
+		return query;
+	}
+	
+	// Transform the GramBooleanQuery tree to Disjunctive normal form (DNF)
+	// which is OR of different ANDs
+	public static GramBooleanQuery toDNF(GramBooleanQuery query) {
```

Comment:
I don't think so, the purpose of it is to transform any boolean expression to DNF. We can change it to package only, since no one else is using it.